File: D05/Rust/src/main.rs
use std::cmp::Ordering;

use itertools::Itertools;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(transparent)]
struct Node(i32);

#[derive(Debug)]
struct Edge(Node, Node);

#[derive(Debug, Default)]
struct Graph(Vec<Edge>);

impl Graph {
    fn ordering(&self, n1: Node, n2: Node) -> Ordering {
        if self.0.iter().any(|Edge(from, to)| *from == n1 && *to == n2) {
            Ordering::Less
        } else if self.0.iter().any(|Edge(from, to)| *from == n2 && *to == n1) {
            Ordering::Greater
        } else {
            Ordering::Equal
        }
    }
}

fn parse(input: &str) -> (Graph, Vec<Vec<Node>>) {
    let mut edges = Vec::new();
    let mut updates = Vec::new();

    for line in input.lines() {
        if line.contains('|') {
            if let Some((left, right)) = line.split_once('|') {
                if let (Ok(a), Ok(b)) = (left.parse::<i32>(), right.parse::<i32>()) {
                    edges.push(Edge(Node(a), Node(b)));
                }
            }
        } else if !line.is_empty() {
            let update = line
                .split(',')
                .filter_map(|x| x.parse::<i32>().map(Node).ok())
                .collect();
            updates.push(update);
        }
    }

    (Graph(edges), updates)
}

fn is_ordered(update: &[Node], graph: &Graph) -> bool {
    update.is_sorted_by(|a, b| graph.ordering(*a, *b) != Ordering::Greater)
}

fn solve_part_one(graph: &Graph, updates: &[Vec<Node>]) -> i32 {
    updates.iter().fold(0, |acc, update| {
        acc + if is_ordered(update, graph) {
            update.get(update.len() / 2).unwrap().0
        } else {
            0
        }
    })
}

fn solve_part_two(graph: &Graph, updates: &[Vec<Node>]) -> i32 {
    updates.iter().fold(0, |acc, update| {
        acc + if !is_ordered(update, graph) {
            update
                .iter()
                .sorted_by(|&&a, &&b| graph.ordering(a, b))
                .nth(update.len() / 2)
                .unwrap()
                .0
        } else {
            0
        }
    })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let data = std::fs::read_to_string("../input.txt")?;

    let (g, t) = parse(&data);

    let p1 = solve_part_one(&g, &t);

    println!("{p1}");

    let p2 = solve_part_two(&g, &t);

    println!("{p2}");

    Ok(())
}

#[cfg(test)]
mod tests {
    static SAMPLE: &str = r"47|53
97|13
97|61
97|47
75|29
61|13
75|53
29|13
97|29
53|29
61|53
97|53
61|29
47|13
75|47
97|75
47|61
75|61
47|29
75|13
53|13

75,47,61,53,29
97,61,53,29,13
75,29,13
75,97,47,61,53
61,13,29
97,13,75,29,47";

    #[test]
    fn test_solve_one() {
        let expected = 143;

        let (g, t) = super::parse(SAMPLE);

        let actual = super::solve_part_one(&g, &t);

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_solve_two() {
        let expected = 123;

        let (g, t) = super::parse(SAMPLE);

        let actual = super::solve_part_two(&g, &t);

        assert_eq!(expected, actual);
    }
}

File: D05/Golang/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Node int
type Edge [2]Node
type Graph struct{ edges []Edge }

type EdgeType int

const (
	In EdgeType = iota
	Out
	None
)

func (g *Graph) AddEdge(edge Edge) {
	g.edges = append(g.edges, edge)
}

func (g *Graph) FindEdgeType(edge Edge) EdgeType {
	for _, e := range g.edges {
		first := e[0]
		second := e[1]
		if edge[0] == first && edge[1] == second {
			return Out
		}
		if edge[0] == second && edge[1] == first {
			return In
		}
	}
	return None
}

func parsePair(input string) (Edge, error) {
	pairs := strings.Split(input, "|")
	if len(pairs) != 2 {
		return Edge{}, errors.New("not a valid pair")
	}
	first, _ := strconv.Atoi(pairs[0])
	second, _ := strconv.Atoi(pairs[1])

	return Edge{Node(first), Node(second)}, nil
}

func parseLine(input string) []Node {
	values := strings.Split(input, ",")
	var vals []Node
	for _, v := range values {
		x, _ := strconv.Atoi(v)
		vals = append(vals, Node(x))
	}
	return vals
}

func parse(input string) (Graph, [][]Node) {
	graph := Graph{}
	var table [][]Node

	for _, line := range strings.Split(strings.Trim(input, "\n"), "\n") {
		edge, err := parsePair(line)
		if err == nil {
			graph.AddEdge(edge)
		} else if len(line) > 0 {
			table = append(table, parseLine(line))
		}

	}

	return graph, table
}

func sortUpdate(update []Node, g *Graph) {
	slices.SortFunc(update, func(n1 Node, n2 Node) int {
		switch g.FindEdgeType(Edge{n1, n2}) {
		case Out:
			return -1
		case None:
			return 0
		default:
			return 1
		}
	})
}

func isOrdered(update []Node, g *Graph) bool {
	for idx := 0; idx < len(update)-1; idx++ {
		first := update[idx]
		second := update[idx+1]
		if g.FindEdgeType(Edge{first, second}) == In {
			return false
		}
	}
	return true
}

func SolvePartOne(graph *Graph, table [][]Node) int {
	var total int
	for _, update := range table {
		if isOrdered(update, graph) {
			total += int(update[len(update)/2])
		}
	}
	return total
}

func SolvePartTwo(graph *Graph, table [][]Node) int {
	var total int
	for _, update := range table {
		if !isOrdered(update, graph) {
			sortUpdate(update, graph)
			total += int(update[len(update)/2])
		}
	}
	return total
}

func main() {
	data, _ := os.ReadFile("../input.txt")

	graph, table := parse(string(data))

	a1 := SolvePartOne(&graph, table)

	fmt.Println(a1)

	a2 := SolvePartTwo(&graph, table)

	fmt.Println(a2)
}
